features/netsocket/EMACMemoryManager.h

```diff
@@ -0,0 +1,159 @@
+/* Copyright (c) 2017 ARM Limited
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#ifndef EMAC_MEMORY_MANAGER_H
+#define EMAC_MEMORY_MANAGER_H
+
+/**
+ * Emac interface memory manager
+ *
+ * This interface provides abstraction for memory modules used in different IP stacks (often to accommodate zero
+ * copy). Emac interface may be required to accept output packets and provide received data using this stack
+ * specific API. This header should be implemented for each IP stack, so that we keep emacs module independent.
+ *
+ * Emac memory interface uses memory buffers to store data. Memory buffers may be either contiguous or chained.
+ * Chaining of the buffers is made using singly linked list.
+ *
+ * Memory buffers can be allocated either from heap or from memory pools. Heap buffers are always contiguous.
+ * Memory pool buffers may be either contiguous or chained depending on allocation size.
```

Comment:
I think we need to distinguish what you call "buffer" here from the "emac_mem_buf_t". I would suggest you use "buffer" and "buffer chain", and say whether calls work on a buffer or an entire chain. Then the EMAC API passes a chain up or down.